### PR TITLE
feature/autohelp overhaul

### LINKED ISSUE #98: Add OCR to autohelp
There are some people who think it is clever to sent errors as screenshots because they don't know how to use haste

### LINKED ISSUE #99: auto-fetch javadoc when there is no tag for an exception

### LINKED ISSUE #100: display more information about an exception in autohelp

### LINKED ISSUE #101: give more information about NPEs in autohelp
give info about what could be null

### LINKED ISSUE #102: show line that caused exception in autohelp

### LINKED ISSUE #103: analyze exception cause as the cause of another exception instead of analyzing it as a standalone exception

### PR DESCRIPTION
**Pull request by DRSchlaubi**
## Bugfixes
This PR changes handling of javadocs
from now on the bot will disable the Javadoc command when it errors while fetching the doc.
I also changed java 9-14 doc to an openjdk hosted version because oracles version is trash

## Autohelp changes
the bot does now also analyze code and highlighted the line that caused the exception according to the stacktrace
if there is no tag for an exception the bot will try to fetch the Javadoc for it

## Todo
.~~- [ ] Handy non-domainish groups ids properly (e.g. okhttp3)~~
- [x] try to find first user element in stack trace

## Github
Fix #103 
Fix #102 
Fix #101
Fix #100 
Fix #99 
Fix #98

### CODE REVIEW

File: src/main/kotlin/com/github/devcordde/devcordbot/commands/general/GoogleCommand.kt [hunk 1]
```diff
@@ -30,7 +30,7 @@ import com.google.api.services.customsearch.Customsearch
 /**
  * Google command.
  */
-class GoogleCommand(private val apiKey: String, private val engineId: String) : AbstractCommand() {
+class GoogleCommand : AbstractCommand() {
 
     private val search =
```
Comment (jkoenig134) [inline src/main/kotlin/com/github/devcordde/devcordbot/commands/general/GoogleCommand.kt]: unused search var

File: src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Brain.kt [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    private fun tryToFindJavadoc(conversation: Conversation) {
+        GlobalScope.launch {
+            val exception = conversation.answer.exception
+            val doc = javadocFinder.findJavadocForClass(exception.exceptionName)
+            if (doc != null) {
+                exception.explanation = exception.explanation ?: doc.description
+                exception.exceptionDoc = doc.uri
+            } else {
+                exception.exceptionDoc = "Ich konnte kein doc finden :C"
```
Comment (jkoenig134) [inline src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Brain.kt]: change to "Es wurde kein Doc gefunden" to be consistent

**Review by jkoenig134 — CHANGES_REQUESTED** [review]

**Review by jkoenig134 — APPROVED** [review]

File: src/main/kotlin/com/github/devcordde/devcordbot/commands/general/_JavadocCommands.kt [hunk 3]
```diff
@@ -18,31 +18,40 @@
...
     }
 
 /**
  * Command for oracle (java 10) doc.
  */
 fun OracleJavaDocCommand(): AbstractCommand = URLJavaDocCommand(
+    "java",
     DocumentedVersion.V_10.url,
     listOf("doc", "docs"),
     "javaodc",
```
Comment (daspositioned) [inline src/main/kotlin/com/github/devcordde/devcordbot/commands/general/_JavadocCommands.kt]: Typo

File: src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Brain.kt [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    private fun buildNpeHint(elements: List<String>, lineNumber: Int): String =
+        """Eines dieser Dinge ${elements.joinToString(
+            prefix = "(`",
+            separator = "`, `",
+            postfix = "`)"
+        )} scheint null zu sein.
+        |Bitte stelle sicher, dass alle davon einen Wert haben oder füge ein `Null-Check` hinzu.
+        |**Tipp**: Füge einen Breakpoint zur zeile `$lineNumber` hinzu um zu checken was genau `null` ist oder benutze die [Helpful NPEs](https://openjdk.java.net/jeps/358) von Java 14+.
```
Comment (daspositioned) [inline src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Brain.kt]: Make "zeile" uppercase

File: src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Conversation.kt [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+     * @property explanation the tag explaining the exception
+     * @property causeClass the class in which the exception is caused
+     * @property causeLine the line in which the exception is caused
+     * @property exceptionDoc a URL to the exceptions javadoc
+     * @property sealed whether this is the final exception data or not
+     */
+    data class ExceptionAnswer(
+        val exceptionName: String,
+        val exceptionMessage: String?,
+        val causee: String?,
```
Comment (daspositioned) [inline src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Conversation.kt:161]: Typo

**Review by daspositioned — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/github/devcordde/devcordbot/commands/general/_JavadocCommands.kt (hunk 3 above)
Comment (DRSchlaubi) [reply src/main/kotlin/com/github/devcordde/devcordbot/commands/general/_JavadocCommands.kt]: wdym?

Comment (daspositioned) [reply src/main/kotlin/com/github/devcordde/devcordbot/commands/general/_JavadocCommands.kt]: "javaodc"

File: src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Conversation.kt (hunk 5 above)
Comment (DRSchlaubi) [reply src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Conversation.kt:161]: the causee is the exception that is caused by the exception and not the exception that causes the causee
similar to the damaged entity being called damagee here:https://hub.spigotmc.org/javadocs/spigot/org/bukkit/event/entity/EntityDamageEvent.html#%3Cinit%3E(org.bukkit.entity.Entity,org.bukkit.event.entity.EntityDamageEvent.DamageCause,double)

File: src/main/kotlin/com/github/devcordde/devcordbot/commands/general/_JavadocCommands.kt (hunk 3 above)
Comment (DRSchlaubi) [reply src/main/kotlin/com/github/devcordde/devcordbot/commands/general/_JavadocCommands.kt]: Yeah reacting with rocket really helped

**Review by daspositioned — APPROVED** [review]

File: src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Conversation.kt (hunk 5 above)
Comment (DRSchlaubi) [reply src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Conversation.kt:161]: Can you also give proper answers or is that no longer cool and you have to react with emojis?

Comment (daspositioned) [reply src/main/kotlin/com/github/devcordde/devcordbot/core/autohelp/Conversation.kt:161]: I am too ashamed to give a proper answer... Good to know

**Pull request merged**